### PR TITLE
VN barcamp : security of WebAdmin

### PR DESCRIPTION
**Pull request by chibenwa**
This is the review for integrating the changes made during the barcamp.

I will take care of review here.

This change set propose : 
 - Reworking WebAdmin configuration
 - Enabling HTTPS (optional)
 - Enabling JWT (optional)
 - Spark upgrade for logging 500 answers
 - Enabling CORS (optional)
 - Metrics for webAdmin (only global for now, not route specific yet)

### CODE REVIEW

**Comment by chibenwa** [comment]
A new one 🥇 
```
[Building Java 8 vn_barcamp_security-7b59c2ee-0cca-4b61-937e-5767a13dc5c1] Location: /org/apache/james/imap/scripts/UidSearchAtomsIndexer.test:2322
[Building Java 8 vn_barcamp_security-7b59c2ee-0cca-4b61-937e-5767a13dc5c1] LastClientMsg: A203 UID SEARCH FROM "Timothy"
[Building Java 8 vn_barcamp_security-7b59c2ee-0cca-4b61-937e-5767a13dc5c1] Expected: '\* SEARCH 1'
[Building Java 8 vn_barcamp_security-7b59c2ee-0cca-4b61-937e-5767a13dc5c1] Actual   : '* SEARCH'
[Building Java 8 vn_barcamp_security-7b59c2ee-0cca-4b61-937e-5767a13dc5c1] 	at org.apache.james.mpt.imapmailbox.suite.UidSearchOnIndex.testSearchAtomsUS(UidSearchOnIndex.java:40)
[Building Java 8 vn_barcamp_security-7b59c2ee-0cca-4b61-937e-5767a13dc5c1] 
[Building Java 8 vn_barcamp_security-7b59c2ee-0cca-4b61-937e-5767a13dc5c1] 
[Building Java 8 vn_barcamp_security-7b59c2ee-0cca-4b61-937e-5767a13dc5c1] Results :
[Building Java 8 vn_barcamp_security-7b59c2ee-0cca-4b61-937e-5767a13dc5c1] 
[Building Java 8 vn_barcamp_security-7b59c2ee-0cca-4b61-937e-5767a13dc5c1] Tests in error: 
[Building Java 8 vn_barcamp_security-7b59c2ee-0cca-4b61-937e-5767a13dc5c1]   UidSearchOnIndex.testSearchAtomsUS:40->AbstractSimpleScriptedTestProtocol.scriptTest:78->AbstractProtocolTestFramework.runSessions:104 » InvalidServerResponse
```

**Comment by chibenwa** [comment]
test this please

File: server/container/guice/guice-common/src/main/java/org/apache/james/modules/server/WebAdminServerModule.java [hunk 1]
```diff
@@ -80,6 +82,26 @@ public WebAdminConfiguration provideWebAdminConfiguration(PropertiesProvider pro
         }
     }
 
+    private HttpsConfiguration readHttpsConfiguration(PropertiesConfiguration configurationFile) {
+        boolean enabled = configurationFile.getBoolean("https.enabled", DEFAULT_HTTPS_DISABLED());
+        if (enabled) {
+            return HttpsConfiguration.builder()
+                .enabled()
+                .raw(configurationFile.getString("https.keystore", null),
```
Comment (quynhn) [inline server/container/guice/guice-common/src/main/java/org/apache/james/modules/server/WebAdminServerModule.java]: Can we extract "null" value as default string value?

File: server/protocols/webadmin/src/main/java/org/apache/james/webadmin/HttpsConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+package org.apache.james.webadmin;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+public class HttpsConfiguration {
```
Comment (mbaechler) [inline server/protocols/webadmin/src/main/java/org/apache/james/webadmin/HttpsConfiguration.java]: looks only related to TLS, not HTTPS (no port)

File: server/protocols/webadmin/src/main/java/org/apache/james/webadmin/HttpsConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        private String keystorePassword;
+        private String truststoreFilePath;
+        private String truststorePassword;
+
+        public Builder enable(boolean isEnabled) {
+            this.enabled = Optional.of(isEnabled);
+            return this;
+        }
+
+        public Builder enabled() {
```
Comment (mbaechler) [inline server/protocols/webadmin/src/main/java/org/apache/james/webadmin/HttpsConfiguration.java]: I would let this property on the parent object

File: server/protocols/webadmin/src/test/java/org/apache/james/webadmin/WebAdminConfigurationTest.java [hunk 4]
```diff
@@ -46,7 +46,7 @@ public void buildShouldWorkWithoutPortWhenDisabled() {
         assertThat(WebAdminConfiguration.builder()
             .disabled()
             .build())
-            .isEqualTo(new WebAdminConfiguration(false, null));
+            .isEqualTo(new WebAdminConfiguration(false, null, HttpsConfiguration.builder().disabled().build()));
```
Comment (mbaechler) [inline server/protocols/webadmin/src/test/java/org/apache/james/webadmin/WebAdminConfigurationTest.java]: that's the point where you need to use extract to check only flags you need to assert

File: server/protocols/jwt/pom.xml [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        <groupId>org.apache.james</groupId>
+        <version>3.0.0-beta6-SNAPSHOT</version>
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>james-server-jwt</artifactId>
+    <packaging>jar</packaging>
+
+    <name>Apache James :: Server :: JWT utils</name>
```
Comment (mbaechler) [inline server/protocols/jwt/pom.xml]: Utils ?

File: server/protocols/webadmin/src/test/java/org/apache/james/webadmin/WebAdminConfigurationTest.java [hunk 6]
```diff
@@ -46,7 +46,11 @@ public void buildShouldWorkWithoutPortWhenDisabled() {
         assertThat(WebAdminConfiguration.builder()
             .disabled()
             .build())
-            .isEqualTo(new WebAdminConfiguration(false, null, HttpsConfiguration.builder().disabled().build()));
+            .isEqualTo(new WebAdminConfiguration(false,
```
Comment (mbaechler) [inline server/protocols/webadmin/src/test/java/org/apache/james/webadmin/WebAdminConfigurationTest.java]: please use extract on assertions

File: server/protocols/jwt/src/test/java/org/apache/james/jwt/JwtTokenVerifierTest.java [hunk 7]
```diff
@@ -131,7 +122,9 @@ public void shouldReturnUserLoginFromValidToken() {
 
     @Test
     public void hasAttributeShouldNotThrowIfClaimValid() throws Exception {
```
Comment (mbaechler) [inline server/protocols/jwt/src/test/java/org/apache/james/jwt/JwtTokenVerifierTest.java]: return true

File: server/protocols/webadmin/src/main/java/org/apache/james/webadmin/HttpsConfiguration.java (hunk 3 above)
Comment (chibenwa) [inline server/protocols/webadmin/src/main/java/org/apache/james/webadmin/HttpsConfiguration.java]: It's still on parent object.

File: server/protocols/webadmin/src/main/java/org/apache/james/webadmin/HttpsConfiguration.java (hunk 2 above)
Comment (chibenwa) [inline server/protocols/webadmin/src/main/java/org/apache/james/webadmin/HttpsConfiguration.java]: Ok

File: server/protocols/webadmin/src/main/java/org/apache/james/webadmin/WebAdminServer.java [hunk 8]
```diff
@@ -41,43 +40,41 @@
...
+        this.configuration = configuration;
         this.routesList = routesList;
-        this.enabled = enabled;
         this.service = Service.ignite();
     }
 
     @VisibleForTesting
     public WebAdminServer(Routes... routes) throws IOException {
-        this(true, new RandomPort(), ImmutableSet.copyOf(routes));
+        this(WebAdminConfiguration.builder()
```
Comment (chibenwa) [inline server/protocols/webadmin/src/main/java/org/apache/james/webadmin/WebAdminServer.java]: More TESTING_CONFIG
Also added a DISABLED_CONFIG

File: server/protocols/webadmin/src/test/java/org/apache/james/webadmin/WebAdminConfigurationTest.java (hunk 6 above)
Comment (chibenwa) [inline server/protocols/webadmin/src/test/java/org/apache/james/webadmin/WebAdminConfigurationTest.java]: Done

File: server/protocols/webadmin/src/test/java/org/apache/james/webadmin/WebAdminConfigurationTest.java (hunk 4 above)
Comment (chibenwa) [inline server/protocols/webadmin/src/test/java/org/apache/james/webadmin/WebAdminConfigurationTest.java]: Done

**Comment by chibenwa** [comment]
test this please

File: server/protocols/webadmin/src/main/java/org/apache/james/webadmin/HttpsConfiguration.java (hunk 3 above)
Comment (mbaechler) [inline server/protocols/webadmin/src/main/java/org/apache/james/webadmin/HttpsConfiguration.java]: ok, I was not clear : enabling TLS should be in Webadmin conf, not in TLS conf

Comment (chibenwa) [inline server/protocols/webadmin/src/main/java/org/apache/james/webadmin/HttpsConfiguration.java]: Ok, but how I do to build an enabled tlsConfig without port? Sorry this does not make any sense to me...

**Comment by chibenwa** [comment]
test this please

**Pull request merged**